clang/lib/Driver/Driver.cpp

```diff
@@ -1290,6 +1292,101 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
+      // Emit a diagnostic if ``--offload-arch`` is invoked without
+      // ``--offload-new driver`` option.
+      if (!C.getInputArgs().hasFlag(options::OPT_offload_new_driver,
+                                    options::OPT_no_offload_new_driver,
+                                    false)) {
+        Diag(clang::diag::err_drv_sycl_offload_arch_new_driver);
+        return;
+      }
+    const ToolChain *HostTC = C.getSingleOffloadToolChain<Action::OFK_Host>();
+    auto AMDTriple = getHIPOffloadTargetTriple(*this, C.getInputArgs());
+    auto NVPTXTriple = getNVIDIAOffloadTargetTriple(*this, C.getInputArgs(),
+                                                    HostTC->getTriple());
+
+    // Attempt to deduce the offloading triple from the set of architectures.
+    // We need to temporarily create these toolchains so that we can access
+    // tools for inferring architectures.
+    llvm::DenseSet<StringRef> Archs;
+    if (NVPTXTriple) {
+      auto TempTC = std::make_unique<toolchains::CudaToolChain>(
+          *this, *NVPTXTriple, *HostTC, C.getInputArgs(), Action::OFK_None);
+      for (StringRef Arch :
+           getOffloadArchs(C, C.getArgs(), Action::OFK_SYCL, &*TempTC, true))
+        Archs.insert(Arch);
+    }
+    if (AMDTriple) {
+      auto TempTC = std::make_unique<toolchains::AMDGPUOpenMPToolChain>(
+          *this, *AMDTriple, *HostTC, C.getInputArgs());
+      for (StringRef Arch :
+           getOffloadArchs(C, C.getArgs(), Action::OFK_SYCL, &*TempTC, true))
+        Archs.insert(Arch);
+    }
+    if (!AMDTriple && !NVPTXTriple) {
+      for (StringRef Arch :
+           getOffloadArchs(C, C.getArgs(), Action::OFK_SYCL, nullptr, true))
+        Archs.insert(Arch);
+    }
+    for (StringRef Arch : Archs) {
+      if (NVPTXTriple && IsSYCLSupportedNVidiaGPUArch(StringToOffloadArchSYCL(
+                             getProcessorFromTargetID(*NVPTXTriple, Arch)))) {
+        DerivedArchs[NVPTXTriple->getTriple()].insert(Arch);
+      } else if (AMDTriple &&
+                 IsSYCLSupportedAMDGPUArch(StringToOffloadArchSYCL(
+                     getProcessorFromTargetID(*AMDTriple, Arch)))) {
+        DerivedArchs[AMDTriple->getTriple()].insert(Arch);
+      } else if (IsSYCLSupportedIntelCPUArch(StringToOffloadArchSYCL(Arch))) {
+        DerivedArchs[MakeSYCLDeviceTriple("spir64_x86_64").getTriple()].insert(
+            Arch);
+      } else if (IsSYCLSupportedIntelGPUArch(StringToOffloadArchSYCL(Arch))) {
+        StringRef IntelGPUArch;
+        IntelGPUArch = mapIntelGPUArchName(Arch).data();
+        DerivedArchs[MakeSYCLDeviceTriple("spir64_gen").getTriple()].insert(
+            IntelGPUArch);
+      } else {
+        Diag(clang::diag::err_drv_invalid_sycl_target) << Arch;
+        return;
+      }
+    }
+    // If the set is empty then we failed to find a native architecture.
+    if (Archs.empty()) {
+      Diag(clang::diag::err_drv_invalid_sycl_target) << "native";
```

Comment:
This error doesn't make any sense to the end user: `SYCL target is invalid: 'native'`.  As we are validating `--offload-arch` values, something that reflects that usage would make more sense.  Can you take advantage of the existing `err_drv_bad_offload_arch_combo` diagnostic?

Also, please add a test for this case.

Reply:
The definition and usage of `err_drv_bad_offload_arch_combo` did not seem appropriate for this use case where we are emitting a diagnostic when `--offload-arch` value is missing/empty.
I added a new diagnostic similar to the existing `err_drv_offload_missing_gpu_arch`

Reply:
There is already `clang::diag::err_drv_failed_to_deduce_target_from_arch` which seems to me reasonable to use, see: https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/Driver.cpp#L1008.